src/Filesystem/Replicate.php

```diff
@@ -214,16 +218,65 @@ public function getAdapterClassNames()
 
     public function createFile($key, Filesystem $filesystem)
     {
-        return $this->primary->createFile($key, $filesystem);
+        if ($this->primary instanceof FileFactory) {
+            return $this->primary->createFile($key, $filesystem);
+        }
+
+        if ($this->secondary instanceof FileFactory) {
+            return $this->secondary->createFile($key, $filesystem);
+        }
+
+        throw new \LogicException(sprintf('None of the adapters implement %s.', FileFactory::class));
+    }
+
+    /**
+     * @param string $key
+     *
+     * @return Stream
+     */
+    public function createStream($key)
+    {
+        if ($this->primary instanceof StreamFactory) {
+            return $this->primary->createStream($key);
+        }
+
+        if ($this->secondary instanceof StreamFactory) {
+            return $this->secondary->createStream($key);
+        }
+
+        throw new \LogicException(sprintf('None of the adapters implement %s.', StreamFactory::class));
     }
 
+    /**
+     * NEXT_MAJOR: Remove this method.
+     *
+     * @deprecated since sonata-project/media-bundle 3.x, use "createStream()" instead.
+     *
+     * @param string $key
+     *
+     * @return Stream
+     */
     public function createFileStream($key, Filesystem $filesystem)
     {
-        return $this->primary->createFileStream($key, $filesystem);
+        @trigger_error(sprintf(
+            'Method "%s()" is deprecated since sonata-project/media-bundle 3.x and will be removed'
+            .' in version 4.0. Use "createStream()" instead.',
+            __METHOD__
+        ), \E_USER_DEPRECATED);
+
+        return $this->createStream($key);
     }
 
     public function listDirectory($directory = '')
```

Comment:
Should we deprecate this method? It is removed in the `master` branch.
It seems to be present only at [`Gaufrette\Adapter\Ftp`](
https://github.com/KnpLabs/Gaufrette/blob/786247eba04d4693e88a80ca9fdabb634675dcac/src/Gaufrette/Adapter/Ftp.php#L230).

Reply:
Maybe we can do the same Logic as the others, if the Ftp is implemented call
it, otherwise exception.

I took the decision of removing it because I didnt found it

Reply:
>Maybe we can do the same Logic as the others, if the Ftp is implemented call
it, otherwise exception.

That was the first solution I tried, but as unlike the other cases, `Ftp` is not an interface, I chose the `method_exists()` approach.
Additionally, this method is currently returning the result of `Ftp::listDirectory()` only for the primary adapter. How should we handle the case for the secondary adapter? At first sight this method doesn't throw an exception or return other value than array.
WDYT?